### PR TITLE
Fix lambda scope, and ghost instance issue

### LINKED ISSUE #476: Production Outage Report: 503 on araya-optinist.com
**Date:** 2026-04-06
**Environment:** Production (subscr-)
**Symptom:** 503 Service Unavailable on https://www.araya-optinist.com/

---

## Summary

The main web service (`subscr-optinist-cloud-service`) has **0 running tasks** because the ECS agent on the sole EC2 instance is stuck in a crash loop, unable to re-register with ECS. The load balancer target group is empty, resulting in 503 errors.

## Timeline

- **2026-04-03 ~11:12 JST** — Last successful deployment of `subscr-optinist-cloud-taskdef:126`
- **2026-04-05 ~23:45 JST** — ECS agent logs show `InvalidInstanceException: Instance 35b0fbc18a774b3bad7b96866cb5596e is invalid` errors begin
- **2026-04-05 23:50:49 JST** — Agent receives termination signal, restarts, but hits terminal error: `Referenced container instance 35b0fbc18a774b3bad7b96866cb5596e not registered`
- **2026-04-05 23:50:51 JST** — Agent enters crash loop: restart -> load stale checkpoint -> fail to register -> exit -> repeat
- **2026-04-06 ~01:00 JST (est.)** — Site begins returning 503 as the ECS-tracked task stops and ALB target group empties

### Primary: `subscr-premium-manager` Lambda Deregistering Non-Premium Instances

**The `cleanup_ghost_ecs_registrations()` function in the `subscr-premium-manager` Lambda caused the outage.** This function runs every 15 minutes via `handle_scheduled_monitoring` and is intended to clean up ghost container instance registrations from stopped/terminated premium instances.

However, it has a critical bug: **it operates on ALL container instances in the cluster, not just premium ones.** The function calls `ecs.list_container_instances(cluster=cluster_name)` with no tier filter, then deregisters any instance where `agentConnected == False` — including the free-tier instance running the main web service.

Other functions in the same Lambda correctly filter for premium instances (e.g., `update_premium_service_desired_count()` uses `filter="attribute:tier == premium"`), but `cleanup_ghost_ecs_registrations()` does not.

**Sequence of events:**
1. The ECS agent on the free-tier instance (`i-0339a1b43ad55ab67`) had a brief connectivity blip
2. `cleanup_ghost_ecs_registrations()` ran, saw `agentConnected=False`
3. Force-deregistered container instance `35b0fbc18a774b3bad7b96866cb5596e` from the cluster
4. The ECS agent reconnected, but its local checkpoint still referenced the deregistered instance ID
5. Agent attempted to re-register using the stale checkpoint and received a terminal error:
   ```
   ClientException: Referenced container instance 35b0fbc18a774b3bad7b96866cb5596e not registered
   ```
6. Agent entered crash loop: restart -> load stale checkpoint -> fail to register -> exit -> repeat
7. No tasks could be placed -> ALB target group emptied -> 503

**Bug location:** `infrastructure/terraform/premium_manager_package/premium_manager.py`, function `cleanup_ghost_ecs_registrations()` (line 5076). The `list_container_instances` call on line 5104 needs a `filter="attribute:tier == premium"` parameter to restrict cleanup to premium instances only.

### Observation: Docker Container Still Running on Old Instance

Despite ECS losing track of the free-tier service, the Docker container for `subscr-optinist-cloud-taskdef:126` was **still running** on `i-0339a1b43ad55ab67` (using 3.27 GB / 6.5 GB memory). The container was started before the deregistration and continued running as an orphan. However, without a valid ECS container instance registration, it was not registered as a target in the ALB target group and was not serving traffic.

## Affected Services

| Service | Desired | Running | Status |
|---------|---------|---------|--------|
| subscr-optinist-cloud-service (main web) | 1 | 0 | **DOWN** - No container instances in capacity provider |
| subscr-background-optinist-cloud-service | 1 | 1 | OK - Running on the existing instance |
| subscr-premium-optinist-cloud-service | 0 | 0 | OK - Intentionally scaled to 0 |

## Infrastructure State

- **Cluster:** `subscr-optinist-cloud-cluster`
- **ASG:** `subscr-optinist-asg` — 1 instance (t3.large) in service, healthy
- **Capacity Provider:** `subscr-optinist-capacity-provider` — managed scaling DISABLED
- **EC2 Instance:** `i-0339a1b43ad55ab67` — running, system checks OK, but ECS agent crash-looping
- **ALB Target Group:** `subscr-optinist-tg` — **empty** (no registered targets)
- **ECS Agent Version:** 1.102.1

## Remediation Steps Taken

1. **Scaled ASG to 2** — A new instance (`i-07584dcec601232c8`) is launching in `ap-northeast-1a`. Once it registers with ECS and the capacity provider, the main service task should be placed on it and the site should recover.

## Recommended Follow-up Actions

1. **Monitor new instance** — Confirm it registers with ECS correctly, the main service task starts, and the ALB target group becomes healthy.

2. **Investigate the old instance** — Once the site is back up:
   - SSH/SSM into `i-0339a1b43ad55ab67`
   - Delete the stale ECS agent checkpoint: `sudo rm /var/lib/ecs/data/agent/data/ecs_agent_data.json`
   - Restart the ECS agent: `sudo systemctl restart ecs`
   - Or simply terminate the instance and scale ASG back to 1

3. **Deploy Lambda fix** — The `cleanup_ghost_ecs_registrations()` function must filter for premium instances only using `filter="attribute:tier == premium"` on the `list_container_instances` call. Fix is in `infrastructure/terraform/premium_manager_package/premium_manager.py`. The same fix should be applied to the `development-premium-manager` Lambda.

4. **Enable deployment circuit breaker** — Currently disabled on the service. Enabling it would allow ECS to automatically roll back failed deployments.

## Key Learnings

- The `cleanup_ghost_ecs_registrations()` function lacked a tier filter, allowing the premium-manager Lambda to deregister non-premium container instances from the shared cluster — this was the direct cause of the outage
- ECS agent checkpoint corruption after forced deregistration causes crash loops with no automatic recovery
- With managed scaling disabled and a single instance, there's no redundancy — a single instance failure takes down the service
- Functions operating on shared clusters must always scope their actions to the correct tier/subset of resources

### PR DESCRIPTION
**Pull request by milesAraya**
#### Summary
- The `cleanup_ghost_ecs_registrations()` function in `premium_manager` caused a production outage (503 on araya-optinist.com) by force-deregistering the free-tier container instance from the shared ECS cluster. The function had no tier filter, so it operated on all instances — not just premium ones.
- Five `list_container_instances` calls across three Lambda packages (`premium_manager`, `premium_cleanup`, `free_manager`) were missing tier-scoped `filter` parameters, allowing each to read or act on container instances belonging to other tiers.
- Additionally, the `agentConnected=False` check in ghost cleanup was too aggressive — a single transient disconnect triggered immediate force-deregistration with no grace period. Added a 5-minute grace period using an EC2 tag to track first-seen disconnect time.

#### Design Decisions
- Used the existing ECS `attribute:tier` filter (already used by `update_premium_service_desired_count`) rather than Python-side filtering, so non-target instances are never returned by the API call in the first place.
- Grace period state is tracked via an EC2 tag (`optinist:agent-disconnected-at`) rather than DynamoDB or SSM Parameter Store. Tags are co-located with the resource, require no additional infrastructure, and are automatically cleaned up when the instance is terminated. The `optinist:` prefix avoids potential conflicts with AWS reserved tag namespaces (e.g., `aws:`, `ecs:`).
- The 5-minute grace period balances safety (transient agent restarts typically recover in seconds) against premium user impact (worst case ~20 minutes before a genuinely stuck agent is cleaned up, given the 15-minute polling interval).
- EC2 stopped/terminated/gone instances are still deregistered immediately — the grace period only applies to running instances with disconnected agents.

### Evidence
- See `REPORT.md` for the full outage investigation, including ECS agent logs, CloudTrail events, and root cause chain.
- ECS agent log from `i-0339a1b43ad55ab67` showing the crash loop after force-deregistration:
  ```
  level=error msg="Unable to register as a container instance with ECS"
    error="ClientException: Referenced container instance 35b0fbc18a774b3bad7b96866cb5596e not registered."
  level=error msg="Received terminal error from RegisterContainerInstance call, exiting"
  ```

### References
- `REPORT.md` — full production outage report (2026-04-06)
- Incident: 503 on https://www.araya-optinist.com/ due to empty ALB target group

#### Files changed
- `infrastructure/terraform/premium_manager_package/premium_manager.py` -- Add `filter="attribute:tier == premium"` to `cleanup_ghost_ecs_registrations()`, `get_ecs_container_instance_id()`, and `cleanup_orphaned_ec2_instances()`. Rework ghost cleanup to use a grace period for agent-disconnected-but-running instances. Handle disconnected container instances with no EC2 ID by deregistering immediately.
- `infrastructure/terraform/premium_cleanup_package/premium_cleanup.py` -- Add `filter="attribute:tier == premium"` to `_get_ecs_container_instance_arn()`.
- `infrastructure/terraform/free_manager_package/free_manager.py` -- Add `filter="attribute:tier == free"` to `get_available_instance_ids()`.
- `infrastructure/terraform/premium_manager.tf` -- Add `ec2:DeleteTags` permission to the Lambda IAM role (required for clearing the grace period tag). Scope `CreateTags`/`DeleteTags` to only the `optinist:agent-disconnected-at` tag key via IAM condition.
- `studio/tests/infrastructure/test_premium_manager.py` -- Add 8 unit tests for `cleanup_ghost_ecs_registrations()` covering tier filter, grace period, no-EC2-ID edge case, and mixed-cluster scenarios.

### Manual Testcases
- Deploy the updated `subscr-premium-manager` Lambda. Verify in CloudWatch Logs that `cleanup_ghost_ecs_registrations` logs "No premium container instances found" (when no premium instances are running) instead of listing free/background instances.
- Start a premium instance, then simulate a brief agent disconnect (e.g., `sudo systemctl stop ecs && sleep 30 && sudo systemctl start ecs`). Verify the instance gets tagged with `optinist:agent-disconnected-at` but is NOT deregistered. Verify the tag is cleared on the next monitoring run after the agent reconnects.
- Stop a premium instance. Verify its container instance is deregistered immediately (no grace period).
- Verify the free-tier and background services remain unaffected by premium ghost cleanup across multiple monitoring cycles.

### Unit, Integration, Contract Test Coverage
- 8 new tests in `TestCleanupGhostECSRegistrations` (63 total pass):
  - `test_uses_premium_tier_filter` — asserts the `filter="attribute:tier == premium"` parameter is passed to the API
  - `test_deregisters_stopped_ec2_immediately` — stopped EC2 is cleaned up with no delay
  - `test_tags_running_instance_on_first_disconnect` — first disconnect only tags, no deregistration
  - `test_skips_within_grace_period` — tagged <5 min ago, left alone
  - `test_deregisters_after_grace_period` — tagged >5 min ago, deregistered
  - `test_clears_tag_on_reconnect` — reconnected agent's tag is removed
  - `test_mixed_cluster_only_deregisters_ghost` — healthy + stopped instance together, only stopped is deregistered
  - `test_deregisters_disconnected_without_ec2_id` — container instance with no EC2 mapping is deregistered immediately
- Run: `pytest studio/tests/infrastructure/test_premium_manager.py -v -- all 63 tests pass`
- Note: Cross-tier isolation is enforced at the AWS API level by the `filter` parameter. Unit tests verify the filter is passed; full cross-tier integration testing requires a real multi-tier cluster (covered by manual testcases above).

#### Difficulties
- Initial investigation was complicated by the background service's container instance (`402578fc`, a t3.micro with 940 MB) being misidentified as a corrupted registration of the free-tier t3.large. Resolved by checking the `ec2InstanceId` and `tier` attribute on the container instance.

#### Risk Assessment
| Area | Risk | Notes |
|------|------|-------|
| premium_manager ghost cleanup | Low | Strictly reduces blast radius — adds tier filter, grace period, and no-EC2-ID handling |
| premium_manager instance lookup | Low | Adds filter to read-only lookup; no behavioral change beyond scoping |
| premium_manager orphan cleanup | Low | Adds filter to ECS query; EC2-side was already premium-filtered |
| premium_cleanup instance lookup | Low | Same pattern as premium_manager lookup fix |
| free_manager instance discovery | Medium | Adds tier filter to rebalancing logic; if free instances lack the `tier=free` ECS attribute, they would be excluded from rebalancing. Verify ECS config sets the attribute on free-tier ASG instances. |
| IAM policy (CreateTags/DeleteTags) | Low | Scoped to only the `optinist:agent-disconnected-at` tag key via `ForAllValues:StringEquals` condition. Instance launch tags use `RunInstances` TagSpecifications which does not require separate `CreateTags` permission. |

### CODE REVIEW

File: infrastructure/terraform/free_manager_package/free_manager.py [hunk 1]
```diff
@@ -927,7 +927,9 @@ def get_available_instance_ids(cluster_name: str, service_name: str) -> List[str
         ]
         for container_status in container_statuses:
```
Comment (itutu-tienday) [inline infrastructure/terraform/free_manager_package/free_manager.py:928]: Regarding this matter, is it necessary to update the documentation (the relevant .md file)?
(Explanation of the mechanism for protecting free instances)

**Comment by milesAraya** [comment]
# Manual Testing Log — Ghost Cleanup & Tier Scoping

**Environment:** development
**Cluster:** development-optinist-cloud-cluster
**Premium instance:** i-05ce937ed998dffdf (container instance d12663a0)
**Date:** 2026-04-06

---

## Pre-flight: Tier filter verification

Confirmed `attribute:tier` filter works on the shared cluster:

| Filter | Result |
|--------|--------|
| `tier == premium` | 1 instance (d12663a0) |
| `tier == free` | 1 instance (bf57b7e2) |
| No filter | 3 instances (includes background service 4e82ac5d, correctly excluded from both tier filters) |

## Pre-flight: IAM tag permissions

Verified `CreateTags` and `DeleteTags` for `optinist:agent-disconnected-at` on the premium instance — both succeeded. Tag was written, confirmed via `describe-tags`, then deleted.

---

## Test 1: Healthy baseline

**Action:** Invoked `development-premium-manager` with scheduled monitoring event.

**Result:** Pass
- Lambda returned `200`, `active_users: 3, running_instances: 1`
- CloudWatch Logs: `"No ghost ECS registrations found"`

## Test 2: Agent disconnect — grace period tagging

**Action:** Stopped ECS agent via SSM (`sudo systemctl stop ecs`), waited 30s for ECS to detect disconnect, invoked monitoring.

**Result:** Pass
- CloudWatch Logs: `"Agent disconnected on i-05ce937ed998dffdf, starting grace period"`
- CloudWatch Logs: `"No ghost ECS registrations found"` (not deregistered)
- EC2 tag applied: `optinist:agent-disconnected-at = 2026-04-06T02:37:22.811037+00:00`

## Test 3: Agent reconnect — tag cleanup

**Action:** Restarted ECS agent via SSM (`sudo systemctl start ecs`), waited 30s, invoked monitoring.

**Result:** Pass
- Agent status: `agentConnected: true`
- EC2 tag `optinist:agent-disconnected-at` cleared (confirmed via `describe-tags`: empty)

File: infrastructure/terraform/free_manager_package/free_manager.py (hunk 1 above)
Comment (milesAraya) [reply infrastructure/terraform/free_manager_package/free_manager.py:928]: No, I don't think so. The fixes were simplifying fixing incomplete guardrails. The type check guardrail already existed in some of the steps in the script, but have now been extended.

**Review by itutu-tienday — APPROVED** [review]
